streams/src/main/java/org/apache/kafka/streams/kstream/Suppress.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressImpl;
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressImpl.BufferConfigImpl;
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressImpl.IntermediateSuppressionImpl;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+import java.time.Duration;
+
+@SuppressWarnings("unused") // complaint about the top-level generic parameters, which actually are necessary
+public interface Suppress<K, V> {
+    interface TimeDefinition<K, V> {
+        long time(ProcessorContext context, K k, V v);
+    }
+
+    enum BufferFullStrategy {
+        EMIT,
+        SPILL_TO_DISK,
+        SHUT_DOWN
+    }
+
+    interface BufferConfig<K, V> {
+        static <K, V> BufferConfig<K, V> withBufferKeys(final long numberOfKeysToRemember) {
+            final BufferConfig<K, V> bufferConfig = new BufferConfigImpl<>();
+            return bufferConfig.bufferKeys(numberOfKeysToRemember);
+        }
+
+        BufferConfig<K, V> bufferKeys(final long numberOfKeysToRemember);
+
+        static <K, V> BufferConfig<K, V> withBufferBytes(final long bytesToUseForSuppressionStorage, final Serializer<K> keySerializer, final Serializer<V> valueSerializer) {
+            final BufferConfig<K, V> bufferConfig = new BufferConfigImpl<>();
+            return bufferConfig.bufferBytes(bytesToUseForSuppressionStorage, keySerializer, valueSerializer);
+        }
+
+        BufferConfig<K, V> bufferBytes(final long bytesToUseForSuppressionStorage, final Serializer<K> keySerializer, final Serializer<V> valueSerializer);
+
+        static <K, V> BufferConfig<K, V> withBufferFullStrategy(final BufferFullStrategy bufferFullStrategy) {
```

Comment:
One meta-comment about the static methods on the interfaces in `Suppress`. Would we want to consider making them `default` instead of `static` in case down the line we want to implement any of these interfaces to have different behavior in the various methods?

Reply:
In general, this is something that's good to start thinking about. I think we have two kinds of interfaces/non-final classes: ones that are for implementing and ones that are for encapsulation.

For implementing: Serde, Windows, StateStore, etc.
For encapsulation: KTable, Materialized, etc.

To me, `Suppress` is in the latter category: since we cast it to `SuppressImpl` immediately, it's not possible to pass in any other implementations of it. The interface/impl split is purely to provide a clean division of external/internal members. The regular Java access modifiers are insufficient for this purpose, since SuppressImpl's internal members need to be accessed from other internal classes, but outside of its package.